src/openrct2/world/tile_inspector.c

```diff
@@ -0,0 +1,339 @@
+#pragma endregion
+
+#include "../game.h"
+#include "../interface/window.h"
+#include "../windows/tile_inspector.h"
+#include "footpath.h"
+#include "map.h"
+#include "tile_inspector.h"
+
+static void map_swap_elements_at(sint32 x, sint32 y, sint16 first, sint16 second)
+{
+	rct_map_element *mapElement = map_get_first_element_at(x, y);
+	rct_map_element *const firstElement = mapElement + first;
+	rct_map_element *const secondElement = mapElement + second;
+
+	// swap_elements shouldn't be called when there is only one element on the tile
+	assert(!map_element_is_last_for_tile(mapElement));
+
+	// Make sure both elements are actually on the current tile
+	sint16 elementCount = 0;
+	do
+	{
+		elementCount++;
+	} while (!map_element_is_last_for_tile(mapElement++));
+	assert(elementCount > max(first, second));
+
+	// Swap their memory
+	rct_map_element temp = *firstElement;
+	*firstElement = *secondElement;
+	*secondElement = temp;
+
+	// Swap the 'last map element for tile' flag if either one of them was last
+	if (map_element_is_last_for_tile(firstElement) || map_element_is_last_for_tile(secondElement))
+	{
+		firstElement->flags ^= MAP_ELEMENT_FLAG_LAST_TILE;
+		secondElement->flags ^= MAP_ELEMENT_FLAG_LAST_TILE;
+	}
+}
+
+/**
+ * Inserts a corrupt element under a given element on a given tile
+ * @param x, y: The coordinates of the tile
+ * @param element_index: The nth element on this tile
+ * Returns 0 on success, MONEY_UNDEFINED otherwise.
+ */
+sint32 tile_inspector_insert_corrupt_at(sint32 x, sint32 y, sint16 element_index, sint32 flags)
+{
+	// Make sure there is enough space for the new element
+	if (!map_check_free_elements_and_reorganise(1))
+	{
+		return MONEY32_UNDEFINED;
+	}
+
+	if (flags & GAME_COMMAND_FLAG_APPLY)
+	{
+		// Create new corrupt element
+		rct_map_element *curruptElement = map_element_insert(x, y, -1, 0); // Ugly hack: -1 guarantees this to be placed first
+		if (curruptElement == NULL)
+		{
+			log_warning("Failed to insert corrupt element.");
+			return MONEY32_UNDEFINED;
+		}
+		curruptElement->type = MAP_ELEMENT_TYPE_CORRUPT;
+
+		// Set the base height to be the same as the selected element
+		rct_map_element *const selectedElement = map_get_first_element_at(x, y) + element_index;
+		curruptElement->base_height = curruptElement->clearance_height = selectedElement->base_height;
+
+		// Move the corrupt element up until the selected list item is reached
+		// this way it's placed under the selected element, even when there are multiple elements with the same base height
+		for (sint16 i = 0; i < element_index; i++)
+		{
+			map_swap_elements_at(x, y, i, i + 1);
+		}
+
+		map_invalidate_tile_full(x << 5, y << 5);
+
+		// Update the tile inspector's list for everyone who has the tile selected
+		rct_window *const tile_inspector_window = window_find_by_class(WC_TILE_INSPECTOR);
+		if (tile_inspector_window != NULL && (uint32)x == windowTileInspectorTileX && (uint32)y == windowTileInspectorTileY)
```

Comment:
Probably, but that would require many new functions to be made. The window can't just simply be invalidated, it needs some logic for the different operations.